Filter test inputs before feeding them to functions under test

1082008 opened the pull request:
This PR filters out noise from test input mentioned in #2

ha5la committed:
Begin implementing input filtering in generator

ha5la committed:
Filter createChallenge and verifyChallenge

ha5la committed:
Filter input of clientVerifyChallenge()

ha5la committed:
Implement filter for clientCreateAuth()

ha5la committed:
Implement filter for serverCreateAccept()

ha5la committed:
Implement filtering clientVerifyAccept()

ha5la committed:
Remove superfluous cloning

ha5la committed:
Implement filtering for serverVerifyAuth()

ha5la committed:
Implement filtering clean()

ha5la committed:
Implement filtering of toKeys()

ha5la committed:
Do some cosmetics

ha5la committed:
Remove unneeded comment

cryptix commented:
Can you also commit the generated `data.json`?

I tried to run `node generate.js` but it hangs somehow after `ok 27 (unnamed assert)`.

* [log.txt](https://github.com/auditdrivencrypto/test-secret-handshake/files/961282/log.txt)

ha5la committed:
Update nodejs version in travis

0.6 and 0.8 is not supported anymore.

1082008 commented:
I could commit data.json but it is not a source file, therefore it shouldn't be committed in the source control system. I will make travis CI to put it in a build artifact but that needs some work.

Until then data.json is available [here](https://github.com/auditdrivencrypto/test-secret-handshake/files/961981/data.json.txt).

cryptix commented:
> it is not a source file

sorry, I don't get that. It is a very static file.

If we don't include it here we have to have nodejs with all the dependencies present in all test environments.

1082008 commented:
It is a very big, *nearly* static file. I see it is easier to include it in source control, but I think we should consider that topic. I open a new issue but commit data.json for now as you requested.

ha5la committed:
Include new version of data.json in source control

Please note that this is generated using a modified clear method which
deletes the nullified properties. This ensures that noise is filtered also
from the output besides the input is clean. There is an open PR for that
on the secret-handshake project.

This is not a source file as it is generated by other source files.
Inclusion of data.json in source control should be reconsidered.
An issue will be opened for this.

1082008 reviewed (commented):
The mentioned PR is: auditdrivencrypto/secret-handshake#9

1082008 commented:
The problem of including `data.json` in source control is moved to #4

1082008 commented on generate.js:
```diff
@@ -3,18 +3,162 @@ var fs = require('fs')
+      args[0] = saved_filtered_input_state // revert the changes fn made on its input
+      output(clone({name: name, args: args, result: result}))
+      args[0] = saved_input_state
+      return fn.apply(null, args)
```
Calling fn twice causes duplicate call of `verifyChallenge()` on client side because `clientVerifyChallenge()` calls `verifyChallenge()`

dominictarr commented:
It's way simpler to just check it in. What we should do, though, is make the tests fully deterministic, (i.e. pass in random seeds for every test), then it will always be the same data.json.

1082008 commented:
I agree, tests should be deterministic. Every time I run generate.js it generates a different data.json. This behavior causes a lot of noise in the diff.
